fl4health/model_bases/apfl_base.py

```diff
@@ -1,19 +1,23 @@
 import copy
-from typing import Dict, List
+from typing import Dict, List, Optional
 
 import torch
 import torch.nn as nn
 
+from fl4health.model_bases.warm_up_base import WarmUpModel
 
-class ApflModule(nn.Module):
+
+class ApflModule(WarmUpModel):
```

Comment:
The way you did this is quite interesting, but I think it might overloading the idea of an APFL module (and other modules like FENDA), especially from the perspective of adaptive alpha calculations etc. The way it is being "loaded" as a warmed up model in the server for the APFL example above is also a bit odd, in the sense that it's not really an APFL model being loaded but a part of one. 

What I was thinking is that we would have a "warm up" loading class that takes in a **source nn.Module**, a **checkpoint** corresponding to a trained version of that module, a **target nn.Module** and a **mapping** that maps the state_dict components of the source module into components of the target module. The class would then perform the transfer and return a "warmed up" model. We could have pre-defined defaults for the various models like APFL/FENDA making assumptions, but the vanilla version would essentially be capable of following any mapping.

For the APFL example, for instance, we'd have a module that simply corresponds to the single model that is duplicated in an APFL module for the source, the target would be an APFL model where we load parameters for `layer_name` to `global_model.layer_name` in the APFL target.

Reply:
That is doable for APFL as they have separate models for personalized section and global one, but I am not sure of same thing in fenda, due to classifier layer. I mean if we train only global module during warmup the dimension of first layer classifier would be half of what it would be in original model and it is a bit tricky how we would load models. Also regardless of that, if we do like that a person needs to define a model for warmup and a model for original training which is a bit more confusing that what do we have now. 
Though I am open to both ways, I just initially thought maybe this way has better interfaca for user.

Reply:
I agree with David that having a warm up loading class is probably the cleanest way to incorporate the warm up functionality. It only requires establishing the loading class and some minimal interventions in the constructor of the BasicClient. I do agree that in some cases it may require a bit more code from the user but in most cases we can use sensible defaults (ie Load all the parameters of the source model into the target model). 

For the case of FENDA, I think you raise a solid point so I talked with David. He mentioned you could just train a global model in the warmup phase and in the subsequent phase discard the classifier and set the global and feature extractor with the remaining parameters. 

I think David is going to speak with you about this further in your meeting, so he can further clarify if I am missing anything.

Reply:
That was the original approach I had for my experimentations, thanks for clarifying. I can do that.